aip/general/0193.md

```diff
@@ -62,11 +62,120 @@ a [`PreconditionFailure`][PreconditionFailure].
 **Note:** `ErrorInfo` represents a special case. There **must** be exactly one
 `ErrorInfo`. It is required.
 
-### Localization
+### Error messages
 
-Error messages **must** be in English. If a localized error message is also
-required, the service **should** use [`google.rpc.LocalizedMessage`][details]
-in the `details` field.
+For each error, the service **must** populate the `message` field on
+[`google.rpc.Status`][Status]. This error message,
+
+- is a developer-facing, human-readable "debug message"
+- is presented in the service's native language
+- both explains the error and offers an actionable resolution to it
+  ([citation](https://cloud.google.com/apis/design/errors#error_model))
+
+**Note:** Sometimes a service will elect to always present
+`Status.message` in English rather than the application's native
+language so that messages are easily searchable in common knowledge
+bases, such as StackOverflow&trade;.
+
+When introducing an error that represents a failure scenario that did
+not previously occur for the service, the payload **must** include
+`ErrorInfo` and any variables found in dynamic segments of the error
+message **must** be present in `ErrorInfo.metadata`. See, "[Dynamic
+variables](#dynamic-variables)".
+
+#### Changing error messages
+
+If your service's API design governing body decides that altering
+`Status.message` represents a backwards incompatible change, then
+`Status.message` **must not** change.
+
+<a name="status-message-warning"></a>
+**Warning:** **Use extreme caution** when changing `Status.message`.
+API consumers often treat this error message as **part of the API
+contract**. Clients perform string matches on the text to differentiate
+one error for another and even parse the error message to extract
+variables from dynamic segments.
+
+There is a safer alternative. The service can chose to include an error
+message by adding [`google.rpc.LocalizedMessage`][LocalizedMessage] to
+[`Status.details`][Status details].
+
+The error message presented in `LocalizedMessage.message` **may** be the
+same as `Status.message` or it **may** be an alternate message.
+
+Reasons to present the same error message in both locations include the
+following:
+
+- The service plans to localize either immediately or in the near
+  future.  See, "[Localization](#localization)".
+- It affords clients the ability to find an error message consistently
+  in one location, `LocalizedMessaage.message`, across all methods of
+  the API Service.
+
+Reasons to present an error message in `LocalizedMessage.message` that
+differs from `Status.message` include the following:
+
+- The service requires an end-user facing error message that differs
+  from the "debug message".
+- Ongoing, iterative error message improvements are expected.
+
+When including `LocalizedMessage`, both fields, `locale` and `message`,
```

Comment:
can you include a proto example here? that'll help make it a little easier to follow the guidance.

Reply:
Do you mean a JSON example? Do you really think that is necessary? The proto has two string fields, `locale` and `message`. `locale` is like "en-US", stated and obvious and `message` has the error message. Also, the proto is linked to more than once.

Reply:
yeah, I guess it's a json payload.

I think an example at the top that has a fully populated response would be great, including:

- status
- code
- errorinfo (domain, reason metadata)
- localizedmessage (locale, message)

So one can see how to properly construct a response and how these values could related to each other.

It's hard to grasp how all of these values fit with each other a visualization to connect them all.

Reply:
Added. PTAL